From 22607c3bea6af006e1e414c970d22afced05402b Mon Sep 17 00:00:00 2001
From: Udita <udidua@habana.ai>
Date: Tue, 06 May 2025 11:09:02 +0300
Subject: [PATCH] [SW-220857] gaudi3: enable pinctrl driver

The patch enables the pinctrl driver.
changes are made in the device tree, Cmake,
Kconfigs and defconfig files for management,
agent, event and pid to support the pinctrl
driver changes in spi and i2c.

Change-Id: Icc2a506267cfc5059047b6c378b218df78640ced
Signed-off-by: Udita <udidua@habana.ai>
---

diff --git a/boards/hl/gaudi3/hl_gaudi3_agent_asic_defconfig b/boards/hl/gaudi3/hl_gaudi3_agent_asic_defconfig
index a1225ec..ffc5c8d 100644
--- a/boards/hl/gaudi3/hl_gaudi3_agent_asic_defconfig
+++ b/boards/hl/gaudi3/hl_gaudi3_agent_asic_defconfig
@@ -50,6 +50,8 @@
 CONFIG_SPI_NOR_SFDP_RUNTIME=y
 CONFIG_FLASH_JESD216_API=y
 CONFIG_SPI_NOR_FLASH_HW_WO_PROTECTION=y
+CONFIG_PINCTRL=y
+CONFIG_PINCTRL_HL_GAUDI3=y
 
 CONFIG_HL_EFUSE=y
 CONFIG_HL_FAKE_EFUSE=y
diff --git a/boards/hl/gaudi3/hl_gaudi3_event_asic_defconfig b/boards/hl/gaudi3/hl_gaudi3_event_asic_defconfig
index ee90c86..867f1f0 100644
--- a/boards/hl/gaudi3/hl_gaudi3_event_asic_defconfig
+++ b/boards/hl/gaudi3/hl_gaudi3_event_asic_defconfig
@@ -52,3 +52,5 @@
 CONFIG_CRC=y
 CONFIG_GPIO=y
 CONFIG_ASSERT_VERBOSE=y
+CONFIG_PINCTRL=y
+CONFIG_PINCTRL_HL_GAUDI3=y
\ No newline at end of file
diff --git a/boards/hl/gaudi3/hl_gaudi3_mgmt_asic_defconfig b/boards/hl/gaudi3/hl_gaudi3_mgmt_asic_defconfig
index 199432a..1d2470f 100644
--- a/boards/hl/gaudi3/hl_gaudi3_mgmt_asic_defconfig
+++ b/boards/hl/gaudi3/hl_gaudi3_mgmt_asic_defconfig
@@ -164,3 +164,5 @@
 CONFIG_LFS_LOG_FILES_LIMIT=10
 CONFIG_LFS_FORMAT_IF_MOUNT_FAILED=n
 CONFIG_ASSERT_VERBOSE=y
+CONFIG_PINCTRL=y
+CONFIG_PINCTRL_HL_GAUDI3=y
diff --git a/boards/hl/gaudi3/hl_gaudi3_pid_asic_defconfig b/boards/hl/gaudi3/hl_gaudi3_pid_asic_defconfig
index e7e0081..76e5585 100644
--- a/boards/hl/gaudi3/hl_gaudi3_pid_asic_defconfig
+++ b/boards/hl/gaudi3/hl_gaudi3_pid_asic_defconfig
@@ -56,3 +56,5 @@
 CONFIG_FLASH_JESD216_API=y
 CONFIG_CRC=y
 CONFIG_ASSERT_VERBOSE=y
+CONFIG_PINCTRL=y
+CONFIG_PINCTRL_HL_GAUDI3=y
diff --git a/boards/hl/gaudi3/hl_gaudi3_preboot_asic_defconfig b/boards/hl/gaudi3/hl_gaudi3_preboot_asic_defconfig
index fbb83bf..4f3144a 100644
--- a/boards/hl/gaudi3/hl_gaudi3_preboot_asic_defconfig
+++ b/boards/hl/gaudi3/hl_gaudi3_preboot_asic_defconfig
@@ -163,3 +163,5 @@
 CONFIG_LFS_LOG_FILES_LIMIT=10
 CONFIG_LFS_FORMAT_IF_MOUNT_FAILED=n
 CONFIG_ASSERT_VERBOSE=y
+CONFIG_PINCTRL=y
+CONFIG_PINCTRL_HL_GAUDI3=y
diff --git a/boards/hl/gaudi3/hl_gaudi3_prebootd1_asic_defconfig b/boards/hl/gaudi3/hl_gaudi3_prebootd1_asic_defconfig
index 275078a..5f3d972 100644
--- a/boards/hl/gaudi3/hl_gaudi3_prebootd1_asic_defconfig
+++ b/boards/hl/gaudi3/hl_gaudi3_prebootd1_asic_defconfig
@@ -18,4 +18,6 @@
 CONFIG_CLOCK_CONTROL_HL_PLL_V2=y
 CONFIG_LINKER_ORPHAN_SECTION_PLACE=y
 CONFIG_SPI_NOR_SFDP_RUNTIME=y
-CONFIG_FLASH_JESD216_API=y
\ No newline at end of file
+CONFIG_FLASH_JESD216_API=y
+CONFIG_PINCTRL=y
+CONFIG_PINCTRL_HL_GAUDI3=y
\ No newline at end of file
diff --git a/boards/hl/gaudi3/pinmux_agent.c b/boards/hl/gaudi3/pinmux_agent.c
index 29434c6..42f9c06 100644
--- a/boards/hl/gaudi3/pinmux_agent.c
+++ b/boards/hl/gaudi3/pinmux_agent.c
@@ -14,22 +14,6 @@
 	set_gpio_pin_func(GPIO_FUNC(1), GPIO_UART0_RXD);
 	set_gpio_pin_func(GPIO_FUNC(1), GPIO_UART0_TXD);
 
-	/* I2C Master 0 and 1 */
-	set_gpio_pin_func(GPIO_FUNC(1), GPIO_I2C_M0_SCL);
-	set_gpio_pin_func(GPIO_FUNC(1), GPIO_I2C_M0_SDA);
-	set_gpio_pin_func(GPIO_FUNC(1), GPIO_I2C_M1_SCL);
-	set_gpio_pin_func(GPIO_FUNC(1), GPIO_I2C_M1_SDA);
-
-	/* SPI 0 and 1 */
-	set_gpio_pin_func(GPIO_FUNC(0), GPIO_SPI_0_MISO);
-	set_gpio_pin_func(GPIO_FUNC(0), GPIO_SPI_0_MOSI);
-	set_gpio_pin_func(GPIO_FUNC(0), GPIO_SPI_0_SCLK_OUT);
-	set_gpio_pin_func(GPIO_FUNC(1), GPIO_SPI_0_CS0); /* cs used as gpio */
-	set_gpio_pin_func(GPIO_FUNC(1), GPIO_SPI_1_MISO);
-	set_gpio_pin_func(GPIO_FUNC(1), GPIO_SPI_1_MOSI);
-	set_gpio_pin_func(GPIO_FUNC(1), GPIO_SPI_1_SCLK_OUT);
-	set_gpio_pin_func(GPIO_FUNC(0), GPIO_SPI_1_CS0); /* cs used as gpio */
-
 	set_gpio_pin_func(GPIO_FUNC(0), GPIO_FUSE_BURN);
 
 	if (!pinmux_is_dual_die())
diff --git a/boards/hl/gaudi3/pinmux_mgmt.c b/boards/hl/gaudi3/pinmux_mgmt.c
index 8c0f92c..607294b 100644
--- a/boards/hl/gaudi3/pinmux_mgmt.c
+++ b/boards/hl/gaudi3/pinmux_mgmt.c
@@ -15,21 +15,6 @@
 	set_gpio_pin_func(GPIO_FUNC(1), GPIO_UART0_RXD);
 	set_gpio_pin_func(GPIO_FUNC(1), GPIO_UART0_TXD);
 
-	/* I2C Bus alert */
-	set_gpio_pin_func(GPIO_FUNC(1), GPIO_SMBUS_MASTER0_ALERT);
-	set_gpio_pin_func(GPIO_FUNC(1), GPIO_SMBUS_MASTER1_ALERT);
-	set_gpio_pin_func(GPIO_FUNC(1), GPIO_SMB_SLAVE_ALERT);
-
-	/* SPI 0 and 1 */
-	set_gpio_pin_func(GPIO_FUNC(0), GPIO_SPI_0_MISO);
-	set_gpio_pin_func(GPIO_FUNC(0), GPIO_SPI_0_MOSI);
-	set_gpio_pin_func(GPIO_FUNC(0), GPIO_SPI_0_SCLK_OUT);
-	set_gpio_pin_func(GPIO_FUNC(1), GPIO_SPI_0_CS0); /* cs used as gpio */
-	set_gpio_pin_func(GPIO_FUNC(1), GPIO_SPI_1_MISO);
-	set_gpio_pin_func(GPIO_FUNC(1), GPIO_SPI_1_MOSI);
-	set_gpio_pin_func(GPIO_FUNC(1), GPIO_SPI_1_SCLK_OUT);
-	set_gpio_pin_func(GPIO_FUNC(0), GPIO_SPI_1_CS0); /* cs used as gpio */
-
 	/* PLL */
 	set_gpio_pin_func(GPIO_FUNC(2), GPIO_PLL_RESET);
 	set_gpio_pin_func(GPIO_FUNC(2), GPIO_PLL_LOCK);
diff --git a/boards/hl/gaudi3/pinmux_preboot.c b/boards/hl/gaudi3/pinmux_preboot.c
index 9bb1b88..0df2b58 100644
--- a/boards/hl/gaudi3/pinmux_preboot.c
+++ b/boards/hl/gaudi3/pinmux_preboot.c
@@ -14,25 +14,7 @@
 	set_gpio_pin_func(GPIO_FUNC(1), GPIO_UART0_RXD);
 	set_gpio_pin_func(GPIO_FUNC(1), GPIO_UART0_TXD);
 
-	/* I2C Master 0 */
-	set_gpio_pin_func(GPIO_FUNC(1), GPIO_I2C_M0_SCL);
-	set_gpio_pin_func(GPIO_FUNC(1), GPIO_I2C_M0_SDA);
-
-	/* I2C Master 1 */
-	set_gpio_pin_func(GPIO_FUNC(1), GPIO_I2C_M1_SCL);
-	set_gpio_pin_func(GPIO_FUNC(1), GPIO_I2C_M1_SDA);
-
 	set_gpio_pin_func(GPIO_FUNC(1), GPIO_HBM_PANIC);
-
-	/* SPI 0 and 1 */
-	set_gpio_pin_func(GPIO_FUNC(0), GPIO_SPI_0_MISO);
-	set_gpio_pin_func(GPIO_FUNC(0), GPIO_SPI_0_MOSI);
-	set_gpio_pin_func(GPIO_FUNC(0), GPIO_SPI_0_SCLK_OUT);
-	set_gpio_pin_func(GPIO_FUNC(1), GPIO_SPI_0_CS0); /* cs used as gpio */
-	set_gpio_pin_func(GPIO_FUNC(1), GPIO_SPI_1_MISO);
-	set_gpio_pin_func(GPIO_FUNC(1), GPIO_SPI_1_MOSI);
-	set_gpio_pin_func(GPIO_FUNC(1), GPIO_SPI_1_SCLK_OUT);
-	set_gpio_pin_func(GPIO_FUNC(0), GPIO_SPI_1_CS0); /* cs used as gpio */
 
 	/* FLR notification to slave die and cpld */
 	set_gpio_pin_func(GPIO_FUNC(1), GPIO_FLR);
diff --git a/dts/arc/hl_gaudi3.dtsi b/dts/arc/hl_gaudi3.dtsi
index d3ca328..9b22958 100644
--- a/dts/arc/hl_gaudi3.dtsi
+++ b/dts/arc/hl_gaudi3.dtsi
@@ -9,6 +9,7 @@
 
 #include <dt-bindings/i2c/i2c.h>
 #include <dt-bindings/gpio/gpio.h>
+#include <dt-bindings/gpio/hl-gaudi3-gpio.h>
 #include <gaudi3/gaudi3_mem_addr_dt.h>
 #include <hl_general_oper.h>
 #include <mem.h>
@@ -384,6 +385,66 @@
 		interrupt-parent = <&idu_intc>;
 		ranges;
 
+		pinctrl: pinctrl@fc48b980 {
+			compatible = "hl,gaudi3-pinctrl";
+			reg = <0xfc48b980 0x60>;
+			status = "okay";
+
+			spi0_default: spi0_default{
+				pinmux = < GPIO_SPI_0_MISO
+				 GPIO_SPI_0_MOSI
+				 GPIO_SPI_0_SCLK_OUT
+				 GPIO_SPI_0_CS0
+				 >;
+				func = < GPIO_FUNC(0)
+				 GPIO_FUNC(0)
+				 GPIO_FUNC(0)
+				 GPIO_FUNC(1)>;
+			};
+
+			spi1_default: spi1_default{
+				pinmux = < GPIO_SPI_1_MISO
+				 GPIO_SPI_1_MOSI
+				 GPIO_SPI_1_SCLK_OUT
+				 GPIO_SPI_1_CS0
+				 >;
+				func = < GPIO_FUNC(1)
+				 GPIO_FUNC(1)
+				 GPIO_FUNC(1)
+				 GPIO_FUNC(0)
+				 >;
+			};
+
+			i2c0_default: i2c0_default{
+				pinmux = < GPIO_I2C_M0_SCL
+				 GPIO_I2C_M0_SDA
+				 >;
+				func = < GPIO_FUNC(1)
+				 GPIO_FUNC(1)
+				 >;
+			};
+
+			i2c1_default: i2c1_default{
+				pinmux = < GPIO_I2C_M1_SCL
+				 GPIO_I2C_M1_SDA
+				 >;
+				func = < GPIO_FUNC(1)
+				 GPIO_FUNC(1)
+				 >;
+			};
+
+			i2c2_default: i2c2_default{
+				pinmux = < GPIO_SMBUS_MASTER0_ALERT
+				 GPIO_SMBUS_MASTER1_ALERT
+				 GPIO_SMB_SLAVE_ALERT
+				 >;
+				func = < GPIO_FUNC(1)
+				 GPIO_FUNC(1)
+				 GPIO_FUNC(1)
+				 >;
+			};
+		};
+
 		pinmux: pinmux@fc48b980 {
 			compatible = "hl,gaudi3-pinmux";
 			reg = <0xfc48b980 0x60>;
@@ -436,6 +497,8 @@
 			#address-cells = <1>;
 			#size-cells = <0>;
 			reg = <0xfc480000 0x100>;
+			pinctrl-0 = <&i2c0_default>;
+			pinctrl-names = "default";
 			interrupts = <320 1>;
 			clock-frequency = <I2C_BITRATE_FAST_PLUS>;
 			node_label = "I2C_0";
@@ -450,6 +513,8 @@
 			reg = <0xfc481000 0x100>;
 			interrupts = <321 2>;
 			clock-frequency = <I2C_BITRATE_FAST_PLUS>;
+			pinctrl-0  = <&i2c1_default>;
+			pinctrl-names = "default";
 			node_label = "I2C_1";
 			sda-hold-time-ns = <100>;
 
@@ -548,6 +613,8 @@
 			#size-cells = <0>;
 			reg = <0xfc482000 0x100>;
 			interrupts = <322 1>;
+			pinctrl-0 = <&i2c2_default>;
+			pinctrl-names = "default";
 			clock-frequency = <I2C_BITRATE_FAST_PLUS>;
 			node_label = "I2C_2";
 			slave_soc0: slave-soc@67 {
@@ -634,6 +701,8 @@
 			reg = <0xfc484000  0x100>;
 			clocks = <&refclk>;
 			interrupts = <324 2>;
+			pinctrl-0 = <&spi0_default>;
+			pinctrl-names = "default";
 			cs-gpios = <&gpio0 8 GPIO_ACTIVE_LOW>;
 			fifo-depth = <16>;
 			max-xfer-size = <32>;
@@ -733,6 +802,8 @@
 			reg = <0xfc483000 0x100>;
 			clocks = <&refclk>;
 			interrupts = <323 1>;
+			pinctrl-0 = <&spi1_default>;
+			pinctrl-names = "default";
 			fifo-depth = <16>;
 			max-xfer-size = <32>;
 			cs-gpios = <&gpio1 2 GPIO_ACTIVE_LOW>;
